intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/method/KotlinMethods.kt

```diff
@@ -0,0 +1,137 @@
+   *     ARETURN
+   *    L2
+   *     LOCALVARIABLE this Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI; L0 L2 0
+   *     LOCALVARIABLE anInternalType Linternal/defaultMethod/AnInternalType; L0 L2 1
+   *     LOCALVARIABLE s Ljava/lang/String; L0 L2 2
+   *     MAXSTACK = 3
+   *     MAXLOCALS = 3
+   * ```
+   *
+   * @see [`JvmDefault` annotation](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/jvm/src/kotlin/jvm/JvmDefault.kt)
+   */
+  fun Method.isKotlinDefaultMethod(): Boolean {
+    val method: Method = this
+    return cache.computeIfAbsent(method.location) {
+      return@computeIfAbsent isKotlinMethodInvokingDefaultImpls(method)
+    }
+  }
+
+  private fun Method.isKotlinMethodInvokingDefaultImpls(method: Method): Boolean {
+    // filter non kotlin classes
+    if (!method.containingClassFile.annotations.any { it.desc == "Lkotlin/Metadata;" }) {
+      return false
+    }
+
+    // Sanity check : if the method does not have bytecode
+    // this heuristic cannot run
+    if (instructions.isEmpty()) {
+      return false
+    }
+
+    // skip non opcodes, and kotlin intrinsics on arguments
+    val candidateOpcodes = mutableListOf<AbstractInsnNode>()
+    var i = 0
+    do {
+      val currentInsnNode = instructions[i]
+      if (currentInsnNode.opcode == -1) {
+        continue
+      }
+
+      // Drop kotlin Intrinsics
+      if (currentInsnNode.opcode == Opcodes.ALOAD
+        && currentInsnNode.next?.opcode == Opcodes.LDC
+        && (currentInsnNode.next?.next?.opcode == Opcodes.INVOKESTATIC
+          && (currentInsnNode.next?.next as MethodInsnNode).owner == "kotlin/jvm/internal/Intrinsics")) {
+        i += 2
+        continue
+      }
+
+      candidateOpcodes.add(currentInsnNode)
+    } while (++i < instructions.size)
+
+
+    val expectedOpcodes = (
+      3 // aload this + invokestatic + (return or areturn)
+        + method.methodParameters.size // aload for each parameter
+      )
+
+    if (candidateOpcodes.size != expectedOpcodes
+      || candidateOpcodes[0].opcode != Opcodes.ALOAD // aload this
+      || candidateOpcodes.slice(1..method.methodParameters.size).any() { it.opcode != Opcodes.ALOAD } // parameters
+      || candidateOpcodes[candidateOpcodes.lastIndex - 1].opcode != Opcodes.INVOKESTATIC
+      || candidateOpcodes.last().opcode !in intArrayOf(
+        Opcodes.RETURN,
+        Opcodes.ARETURN,
+        Opcodes.DRETURN,
+        Opcodes.FRETURN,
+        Opcodes.IRETURN,
+        Opcodes.LRETURN
+      )
+    ) {
+      return false
+    }
+
+    val methodInsnNode = candidateOpcodes[candidateOpcodes.lastIndex - 1] as MethodInsnNode
+    if (methodInsnNode.name != method.name || !methodInsnNode.owner.endsWith("\$DefaultImpls")) {
+      return false
+    }
+
+    val actualKotlinOwner = methodInsnNode.owner.substring(0, methodInsnNode.owner.length - "\$DefaultImpls".length)
+
+    // do we want to walk the whole class hierarchy?
```

Comment:
It's not necessary because these methods always delegate to the immediate superinterface. E.g.
```
interface A {
    fun foo() {}
}

interface B : A

class C : B
```
`C.foo` will have a call to `B$DefaultImpls.foo`.

Reply:
Thank you I will update the comment.